### DISCUSSION TITLE
picocli.shell.jline3.PicocliCommands and FilesCompleter

### DISCUSSION THREAD

**Discussion by rcauble** [remkop/picocli#1790] [Q&A]
Hi, I am experimenting with picocli.shell.jline3.PicocliCommands and I was trying to figure out the best way to get it to tab complete for options that are expected to be a local file. 

I've done something like this: 

private static class MyCommands extends PicocliCommands {
        private final Supplier<Path> workDir;

        public MyCommands(CommandLine cmd, Supplier<Path> workDir) {
            super(cmd);
            this.workDir = workDir;
        }

        public SystemCompleter compileCompleters() {
            SystemCompleter rv = super.compileCompleters();
            rv.add(Arrays.asList("mycommand1","mycommand2"),
                    new Completers.FilesCompleter(workDir));
            return rv;
        }
    }
}

It seems to generally work, but file completion ends up happening in all contexts within "mycommand1" and "mycommand2" rather than just in contexts where I've specified a File. Was wondering if there's something simple I'm missing. Ideally if I could specify the completer via annotation on specific options? So something like:

@CommandLine.Option(names = "-myoption",
       completer=FilesCompleter.class)
private File value;

**Comment by remkop** [comment]
The SystemCompleter stuff is a JLine API, so there is no connection with specific picocli options. You can add generic completers, or completers for JLine commands.

The picocli API actually provides an API that is very similar to what you mentioned as "Ideally if I could specify the completer via annotation on specific options?"

The API that picocli provides uses the `completionCandidates` attribute:

```java
@CommandLine.Option(names = "-myoption",
        completionCandidates = FilesCompleter.class)
File f;
```

So, picocli can already do this [with the completionCandidates attribute](https://picocli.info/#_completion_candidates_variable) of the `@Option` annotation.
Hopefully that meets your needs.

**Reply by rcauble** [reply]
Thanks, I gave the following a try:

```
class FileCompleter implements Iterable<String> {
    @NotNull
    @Override
    public Iterator<String> iterator() {
        List<String> files = new ArrayList<>();
        recurse(Path.of(System.getProperty("user.dir")).toFile(), "./", files);
        return files.iterator();
    }

    private void recurse(File root, String path, List<String> rv) {
        File[] files =
                root.listFiles();
        if (files == null) {
            return;
        }
        for (File file : files) {
            String name =
                    file.isDirectory() ? file.getName() + "/" : file.getName();
            rv.add(path+name);
            recurse(file, path + name, rv);
        }
    }
}
```
Unfortunately it ends up returning too many completions to be useable (hundreds). Iterable<String> as an interface doesn't have way to filter based on what the user has typed.
